---
title: "generate_phers"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{generate_phers}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = '#>',
  retina = 2)

foreach::registerDoSEQ()
```

```{r setup}
library('phers')
library('data.table')
library('iterators')
library('ggplot2')
library('BEDMatrix')

# doParallel::registerDoParallel() # register a parallel backend to minimize runtime
```

```{r}
# map ICD codes to phecodes
phecodeOccurrences = getPhecodeOccurrences(icdSample)

# calculate weights
weights = getWeights(demoSample, phecodeOccurrences)

# vector of OMIM disease IDs to calculate PheRS for
diseaseIds = 154700

# mape diseases to phecodes
diseasePhecodeMap = mapDiseaseToPhecode()

# calculate PheRS
scores = getScores(
  demoSample, phecodeOccurrences, weights, diseasePhecodeMap[disease_id %in% diseaseIds])
 
# calculate residual PheRS
rscores = getResidualScores(demoSample, scores, glmFormula = as.formula(~ sex))

# calculate PheRS using pre-calculated weights
# Recommended when data provided by the user has low sample size
pWeights =  preCalcWeights[phecode %in% unique(phecodeOccurrences$phecode)]
scores = getScores(
  demoSample, phecodeOccurrences, pWeights, diseasePhecodeMap[disease_id == diseaseIds])
```


perform genotype association 
```{r}
# create a map of diseases and variants
diseaseId = 154700
diseaseGeneMap = data.table(disease_id = diseaseId, gene = 'FBN1')
geneVarMap = data.table(gene = 'FBN1', vid = paste0('snp', 1:10))
diseaseVariantMap = merge(diseaseGeneMap, geneVarMap, by = 'gene')

npop = 50
nvar = 10
genoSample = BEDMatrix(system.file('extdata', 'geno_sample.bed', package = 'phers'))
colnames(genoSample) = paste0('snp', 1:nvar)
rownames(genoSample) = 1:npop


# run genetic association tests
genoStats = getGeneticAssociations(
  scores, genoSample, demoSample, diseaseVariantMap, glmFormula = as.formula(~ sex), modelType = 'additive')

genoStats = merge(geneVarMap[, .(gene, vid)], genoStats, by = 'vid')

genoStats
```


run positive control analysis
```{r}
dxStatus = getDxStatus(demoSample, icdSample)

plotInput = merge(rscores, dxStatus, by = c('person_id', 'disease_id'))

ggplot(plotInput[disease_id %in% diseaseIds]) + geom_boxplot(
  aes(x = factor(dx_status), y = resid_score, color = factor(dx_status)))
```
