<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Perform association tests between phenotype risk scores and genotypes — getGeneticAssociations • phers</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform association tests between phenotype risk scores and genotypes — getGeneticAssociations"><meta name="description" content="The association test for each disease-variant pair is based on a linear
model, with the phenotype risk score as the dependent variable."><meta property="og:description" content="The association test for each disease-variant pair is based on a linear
model, with the phenotype risk score as the dependent variable."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phers</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://www.hugheylab.org/software/" target="_blank" aria-label="Hughey Lab">Hughey Lab</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/hugheylab/phers/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform association tests between phenotype risk scores and genotypes</h1>
      <small class="dont-index">Source: <a href="https://github.com/hugheylab/phers/blob/main/R/genetic_association.R" class="external-link"><code>R/genetic_association.R</code></a></small>
      <div class="d-none name"><code>getGeneticAssociations.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The association test for each disease-variant pair is based on a linear
model, with the phenotype risk score as the dependent variable.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getGeneticAssociations</span><span class="op">(</span></span>
<span>  <span class="va">scores</span>,</span>
<span>  <span class="va">genotypes</span>,</span>
<span>  <span class="va">demos</span>,</span>
<span>  <span class="va">diseaseVariantMap</span>,</span>
<span>  <span class="va">lmFormula</span>,</span>
<span>  modelType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genotypic"</span>, <span class="st">"additive"</span>, <span class="st">"dominant"</span>, <span class="st">"recessive"</span><span class="op">)</span>,</span>
<span>  level <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  dopar <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-scores">scores<a class="anchor" aria-label="anchor" href="#arg-scores"></a></dt>
<dd><p>A data.table of phenotype risk scores. Must have columns
<code>person_id</code>, <code>disease_id</code>, <code>score</code>.</p></dd>


<dt id="arg-genotypes">genotypes<a class="anchor" aria-label="anchor" href="#arg-genotypes"></a></dt>
<dd><p>A matrix or 'BEDMatrix' object containing genetic data, with
rownames corresponding to <code>person_id</code>s in <code>demos</code> and <code>scores</code>, and
colnames corresponding to <code>variant_id</code>s in <code>diseaseVariantMap</code>.</p></dd>


<dt id="arg-demos">demos<a class="anchor" aria-label="anchor" href="#arg-demos"></a></dt>
<dd><p>A data.table of characteristics for each person in the cohort.
Must have column <code>person_id</code>.</p></dd>


<dt id="arg-diseasevariantmap">diseaseVariantMap<a class="anchor" aria-label="anchor" href="#arg-diseasevariantmap"></a></dt>
<dd><p>A data.table indicating which genetic variants to
test for association with phenotype risk scores for which diseases. Must
have columns <code>disease_id</code> and <code>variant_id</code>.</p></dd>


<dt id="arg-lmformula">lmFormula<a class="anchor" aria-label="anchor" href="#arg-lmformula"></a></dt>
<dd><p>A formula representing the linear model (excluding the term
for genotype) to use for the association tests. All terms in the formula
must correspond to columns in <code>demos</code>.</p></dd>


<dt id="arg-modeltype">modelType<a class="anchor" aria-label="anchor" href="#arg-modeltype"></a></dt>
<dd><p>A string indicating how to encode genotype in the model.</p></dd>


<dt id="arg-level">level<a class="anchor" aria-label="anchor" href="#arg-level"></a></dt>
<dd><p>A number indicating the level of the confidence interval.
Default is 0.95.</p></dd>


<dt id="arg-dopar">dopar<a class="anchor" aria-label="anchor" href="#arg-dopar"></a></dt>
<dd><p>Logical indicating whether to run calculations in parallel if
a parallel backend is already set up, e.g., using
<code><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">doParallel::registerDoParallel()</a></code>. Recommended to minimize runtime.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.table of statistics for the association tests (if a model
fails to converge, NAs will be reported):</p><ul><li><p><code>disease_id</code>: Disease identifier</p></li>
<li><p><code>variant_id</code>: Variant identifier</p></li>
<li><p><code>n_total</code>: Number of persons with non-missing genotype data for the
given variant.</p></li>
<li><p><code>n_wt</code>: Number of persons homozygous for the wild-type allele.</p></li>
<li><p><code>n_het</code>: Number of persons having one copy of the alternate allele.</p></li>
<li><p><code>n_hom</code>: Number of persons homozygous for the alternate allele.</p></li>
<li><p><code>beta</code>: Coefficient for the association of genotype with score</p></li>
<li><p><code>se</code>: Standard error for <code>beta</code></p></li>
<li><p><code>pval</code>: P-value for <code>beta</code> being non-zero</p></li>
<li><p><code>ci_lower</code>: Lower bound of the confidence interval for <code>beta</code></p></li>
<li><p><code>ci_upper</code>: Upper bound of the confidence interval for <code>beta</code></p></li>
</ul><p>If <code>modelType</code> is "genotypic", the data.table will include separate
statistics for heterozygous and homozygous genotypes.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/r/stats/lm.html" class="external-link">stats::lm()</a></code>, <code><a href="https://rdrr.io/r/stats/confint.html" class="external-link">stats::confint()</a></code>, <code><a href="getScores.html">getScores()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://r-datatable.com" class="external-link">'data.table'</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/QuantGen/BEDMatrix" class="external-link">'BEDMatrix'</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># map ICD codes to phecodes</span></span></span>
<span class="r-in"><span><span class="va">phecodeOccurrences</span> <span class="op">=</span> <span class="fu"><a href="getPhecodeOccurrences.html">getPhecodeOccurrences</a></span><span class="op">(</span><span class="va">icdSample</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># calculate weights</span></span></span>
<span class="r-in"><span><span class="va">weights</span> <span class="op">=</span> <span class="fu"><a href="getWeights.html">getWeights</a></span><span class="op">(</span><span class="va">demoSample</span>, <span class="va">phecodeOccurrences</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># OMIM disease IDs for which to calculate phenotype risk scores</span></span></span>
<span class="r-in"><span><span class="va">diseaseId</span> <span class="op">=</span> <span class="fl">154700</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># map diseases to phecodes</span></span></span>
<span class="r-in"><span><span class="va">diseasePhecodeMap</span> <span class="op">=</span> <span class="fu"><a href="mapDiseaseToPhecode.html">mapDiseaseToPhecode</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># calculate scores</span></span></span>
<span class="r-in"><span><span class="va">scores</span> <span class="op">=</span> <span class="fu"><a href="getScores.html">getScores</a></span><span class="op">(</span><span class="va">weights</span>, <span class="va">diseasePhecodeMap</span><span class="op">[</span><span class="va">disease_id</span> <span class="op">==</span> <span class="va">diseaseId</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># map diseases to genetic variants</span></span></span>
<span class="r-in"><span><span class="va">nvar</span> <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="va">diseaseVariantMap</span> <span class="op">=</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></span><span class="op">(</span>disease_id <span class="op">=</span> <span class="va">diseaseId</span>, variant_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'snp'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">nvar</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load sample genetic data</span></span></span>
<span class="r-in"><span><span class="va">npop</span> <span class="op">=</span> <span class="fl">50</span></span></span>
<span class="r-in"><span><span class="va">genoSample</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BEDMatrix/man/BEDMatrix.html" class="external-link">BEDMatrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'geno_sample.bed'</span>, package <span class="op">=</span> <span class="st">'phers'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting number of samples and rownames from geno_sample.fam...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting number of variants and colnames from geno_sample.bim...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">genoSample</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'snp'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">nvar</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">genoSample</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">npop</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run genetic association tests</span></span></span>
<span class="r-in"><span><span class="va">genoStats</span> <span class="op">=</span> <span class="fu">getGeneticAssociations</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">scores</span>, <span class="va">genoSample</span>, <span class="va">demoSample</span>, <span class="va">diseaseVariantMap</span>, lmFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span>,</span></span>
<span class="r-in"><span>  modelType <span class="op">=</span> <span class="st">'additive'</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jake Hughey, Layla Aref.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

