<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform association tests between phenotype risk scores and genotypes — getGeneticAssociations • phers</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform association tests between phenotype risk scores and genotypes — getGeneticAssociations"><meta property="og:description" content="The association test for each disease-variant pair is based on a
generalized linear model, with the phenotype risk score as the dependent
variable."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">phers</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/hugheylab/phers/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform association tests between phenotype risk scores and genotypes</h1>
    <small class="dont-index">Source: <a href="https://github.com/hugheylab/phers/blob/HEAD/R/genetic_association.R" class="external-link"><code>R/genetic_association.R</code></a></small>
    <div class="hidden name"><code>getGeneticAssociations.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The association test for each disease-variant pair is based on a
generalized linear model, with the phenotype risk score as the dependent
variable.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">getGeneticAssociations</span><span class="op">(</span>
  <span class="va">scores</span>,
  <span class="va">genotypes</span>,
  <span class="va">demos</span>,
  <span class="va">diseaseVariantMap</span>,
  <span class="va">glmFormula</span>,
  modelType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genotypic"</span>, <span class="st">"additive"</span>, <span class="st">"dominant"</span>, <span class="st">"recessive"</span><span class="op">)</span>,
  level <span class="op">=</span> <span class="fl">0.95</span>,
  dopar <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>scores</dt>
<dd><p>A data.table of phenotype risk scores. Must have columns
<code>person_id</code>, <code>disease_id</code>, <code>score</code>.</p></dd>
<dt>genotypes</dt>
<dd><p>A matrix or 'BEDMatrix' object containing genetic data, with
rownames corresponding to <code>person_id</code>s in <code>demos</code> and <code>scores</code>, and
colnames corresponding to <code>vid</code>s in <code>diseaseVariantMap</code>.</p></dd>
<dt>demos</dt>
<dd><p>A data.table of characteristics for each person in the cohort.
Must have column <code>person_id</code>.</p></dd>
<dt>diseaseVariantMap</dt>
<dd><p>A data.table indicating which genetic variants to
test for association with phenotype risk scores for which diseases. Must
have columns <code>disease_id</code> and <code>vid</code>.</p></dd>
<dt>glmFormula</dt>
<dd><p>A formula representing the linear model (excluding the term
for genotype) to use for the association tests. All terms in the formula
must correspond to columns in <code>demos</code>.</p></dd>
<dt>modelType</dt>
<dd><p>A string indicating how to encode genotype in the model.</p></dd>
<dt>level</dt>
<dd><p>A number indicating the level of the confidence interval.
Default is 0.95.</p></dd>
<dt>dopar</dt>
<dd><p>Logical indicating whether to run calculations in parallel if
a parallel backend is already set up, e.g., using
<code><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">doParallel::registerDoParallel()</a></code>. Recommended to minimize runtime.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data.table of statistics for the association tests (if a model
fails to converge, NAs will be reported):</p><ul><li><p><code>disease_id</code>: Disease identifier</p></li>
<li><p><code>vid</code>: Variant identifier</p></li>
<li><p><code>n_total</code>: Number of persons with non-missing genotype data for the
given variant.</p></li>
<li><p><code>n_wt</code>: Number of persons homozygous for the wild-type allele.</p></li>
<li><p><code>n_het</code>: Number of persons having one copy of the alternate allele.</p></li>
<li><p><code>n_hom</code>: Number of persons homozygous for the alternate allele.</p></li>
<li><p><code>beta</code>: Coefficient for the association of genotype with score</p></li>
<li><p><code>se</code>: Standard error for <code>beta</code></p></li>
<li><p><code>pval</code>: P-value for <code>beta</code> being non-zero</p></li>
<li><p><code>ci_lower</code>: Lower bound of the confidence interval for <code>beta</code></p></li>
<li><p><code>ci_upper</code>: Upper bound of the confidence interval for <code>beta</code></p></li>
</ul><p>If <code>modelType</code> is "genotypic", the data.table will include separate
statistics for heterozygous ("het") and homozygous ("hom") genotypes.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/r/stats/glm.html" class="external-link">stats::glm()</a></code>, <code><a href="https://rdrr.io/r/stats/confint.html" class="external-link">stats::confint()</a></code>, <code><a href="getScores.html">getScores()</a></code>, <code><a href="phers.html">phers()</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://r-datatable.com" class="external-link">'data.table'</a></span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/QuantGen/BEDMatrix" class="external-link">'BEDMatrix'</a></span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># map ICD codes to phecodes</span></span>
<span class="r-in"><span class="va">phecodeOccurrences</span> <span class="op">=</span> <span class="fu"><a href="getPhecodeOccurrences.html">getPhecodeOccurrences</a></span><span class="op">(</span><span class="va">icdSample</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># calculate weights</span></span>
<span class="r-in"><span class="va">weights</span> <span class="op">=</span> <span class="fu"><a href="getWeights.html">getWeights</a></span><span class="op">(</span><span class="va">demoSample</span>, <span class="va">phecodeOccurrences</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># OMIM disease IDs for which to calculate phenotype risk scores</span></span>
<span class="r-in"><span class="va">diseaseId</span> <span class="op">=</span> <span class="fl">154700</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># map diseases to phecodes</span></span>
<span class="r-in"><span class="va">diseasePhecodeMap</span> <span class="op">=</span> <span class="fu"><a href="mapDiseaseToPhecode.html">mapDiseaseToPhecode</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># calculate scores</span></span>
<span class="r-in"><span class="va">scores</span> <span class="op">=</span> <span class="fu"><a href="getScores.html">getScores</a></span><span class="op">(</span></span>
<span class="r-in">  <span class="va">demoSample</span>, <span class="va">phecodeOccurrences</span>, <span class="va">weights</span>, <span class="va">diseasePhecodeMap</span><span class="op">[</span><span class="va">disease_id</span> <span class="op">==</span> <span class="va">diseaseId</span><span class="op">]</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># map diseases to genetic variants</span></span>
<span class="r-in"><span class="va">nvar</span> <span class="op">=</span> <span class="fl">10</span></span>
<span class="r-in"><span class="va">diseaseVariantMap</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html" class="external-link">data.table</a></span><span class="op">(</span>disease_id <span class="op">=</span> <span class="va">diseaseId</span>, vid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'snp'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">nvar</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># load sample genetic data</span></span>
<span class="r-in"><span class="va">npop</span> <span class="op">=</span> <span class="fl">50</span></span>
<span class="r-in"><span class="va">genoSample</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BEDMatrix/man/BEDMatrix.html" class="external-link">BEDMatrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'geno_sample.bed'</span>, package <span class="op">=</span> <span class="st">'phers'</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting number of samples and rownames from geno_sample.fam...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting number of variants and colnames from geno_sample.bim...</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">genoSample</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'snp'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">nvar</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">genoSample</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">npop</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># run genetic association tests</span></span>
<span class="r-in"><span class="va">genoStats</span> <span class="op">=</span> <span class="fu">getGeneticAssociations</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">scores</span>, <span class="va">genoSample</span>, <span class="va">demoSample</span>, <span class="va">diseaseVariantMap</span>, glmFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span>,</span>
<span class="r-in">  modelType <span class="op">=</span> <span class="st">'additive'</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">  </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jake Hughey, Layla Aref.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

